test/integration/ddl_sharding_key_test.lua

```diff
@@ -0,0 +1,181 @@
+local fio = require('fio')
+
+local t = require('luatest')
+
+local helpers = require('test.helper')
+
+local ok = pcall(require, 'ddl')
+if not ok then
+    t.skip('Lua module ddl is required to run test')
+end
+
+local pgroup = t.group('ddl_sharding_key', {
+    {engine = 'memtx'},
+    {engine = 'vinyl'},
+})
+
+pgroup.before_all(function(g)
+    g.cluster = helpers.Cluster:new({
+        datadir = fio.tempdir(),
+        server_command = helpers.entrypoint('srv_ddl'),
+        use_vshard = true,
+        replicasets = helpers.get_test_replicasets(),
+        env = {
+            ['ENGINE'] = g.params.engine,
+        },
+    })
+    g.cluster:start()
+    local result, err = g.cluster.main_server.net_box:eval([[
+        local ddl = require('ddl')
+
+        local ok, err = ddl.get_schema()
+        return ok, err
+    ]])
+    t.assert_equals(type(result), 'table')
+    t.assert_equals(err, nil)
+end)
+
+pgroup.after_all(function(g) helpers.stop_cluster(g.cluster) end)
+
+pgroup.before_each(function(g)
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_name_key')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_name_key_uniq_index')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_name_key_non_uniq_index')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_secondary_idx_name_key')
+end)
+
+pgroup.test_select = function(g)
+    local tuple = {2, box.NULL, 'Ivan', 20}
+
+    -- insert tuple
+    local result, err = g.cluster.main_server.net_box:call('crud.insert', {
+        'customers_name_key', tuple
+    })
+
+    t.assert_equals(err, nil)
+    t.assert_not_equals(result, nil)
+    t.assert_equals(#result.rows, 1)
+
+    local conditions = {{'==', 'id', 2}}
+
+    local result, err = g.cluster.main_server.net_box:call('crud.select', {
+        'customers_name_key', conditions,
+    })
+
+    t.assert_equals(err, nil)
+    t.assert_not_equals(result, nil)
+    t.assert_equals(#result.rows, 1)
+end
+
+-- Right now CRUD's plan for select doesn't support sharding key and it leads
+-- to map reduce (select on all replicasets). To avoid map-reduce one need to
+-- add a separate index by field name, used in select's condition. We plan to
+-- fix this in scope of https://github.com/tarantool/crud/issues/213
```

Comment:
Here I stuck for a long time, but finally understood what you want to explain. Whether the following description is correct?

<details>
<summary>My explanation of commit 'Support custom sharding key in select'</summary>

In some cases `crud.select()` may understand that a request will involve data only from one replicaset and deduce the sharding key from provided conditions. For such requests it calculates bucket id from the sharding key and fetch data only from the replicaset pointed by this bucket id. Otherwise it performs map-reduce (fetches data from all replicasets and merge them).

Let's stop on this a bit: if `crud.select()` performs map-reduce, the request does not depend on the sharding key definition in the database schema (in tarantool/ddl module). However, if it calls only one replicaset, it should be correct one.

This commit eliminates `crud.select()`'s assumption that the sharding key definition is always the same as the primary key definition: so the storage request will not going to a wrong replicaset. However this commit does not refactor deducing of the sharding key, so there are many cases, when it'll fallback to map-reduce — even if enough data is provided in conditions. It'll be resolved in #213.

Strictly speaking, the following cases are handled now using one replicaset call:

* sharding key == primary key, EQ condition is provided for primary index or for all fields of the primary index.
* sharding key != primary key, EQ condition is provided for a secondary index, which is choosen for lookup on storage (or for all field in the index).

Everything else leads to map-reduce. So, say, the following cases will be handled with map-reduce until #213 will be resolved:

* sharding key != primary key, EQ condition is provided for primary index.
* sharding key != primary key, sharding key != sk1, sk1 is choosen for lookup on storage, all values for the sharding key are provided in conditions.
</details>

Information of this kind should be in the commit message (at least in some sketchy form). #213 also requires clarification. Otherwise it is *really* hard to understand what is fixed and what is postponed.

The code looks correct, sure, and we can proceed with the patch. But I spent hours to understand it and formalize my understanding. Accurate description would safe my effort, review iteration time and maybe even amount of review iterations. Even if we don't care of future readers. It would be more cheap in terms of sum of our efforts, if you would describe it, because as author of the code you already know all details.

I guess you overrate my ability to understand code from others. I would prefer if you would think that a reader of your commits is a bit dumb :)

I hope I can give several general suggestions:

* The commit definitely needs some explanation: at least because there is a problem that is resolved here and there is a problem that is postponed. A reader should understand what is the first problem and what is the second.
* We should tend to describe things in terms of crud and general knowledge: say, I can understand 'code that dedudes a value of sharding key from provided conditions misses the following cases', but I was unable to understand 'select now supports custom sharding key, while select plan does not'. How should I understand that 'select plan' means code that deduces a sharding key and how it is connected to decision of performing single storage call or map-reduce? The plan construction involves many other things aside of sharding key determination.
* I'll also highlight that #213 is marked as part of #166, but is not resolved in this PR, which closes #166.

Reply:
Or, say, just a comment for the code that extracts sharding key:

> -- TODO: After enabling support of sharding keys that are not equal to primary keys, we should handle it differently: it is not enough to look just on scan value, we should traverse all conditions. Now missed cases lead to map-reduce. Will be resolved in #213.

...would give me enough understanding of your change and what will be resolved later.

Personally I'm not good in concise explanations, you know. Maybe you'll make it better.

Reply:
Test has been updated.

```diff
diff --git a/test/integration/ddl_sharding_key_test.lua b/test/integration/ddl_sharding_key_test.lua
index a82b0e3..add6eca 100644
--- a/test/integration/ddl_sharding_key_test.lua
+++ b/test/integration/ddl_sharding_key_test.lua
@@ -3,6 +3,7 @@ local crud = require('crud')
 local t = require('luatest')
 
 local helpers = require('test.helper')
+local storage_stat = require('test.helpers.storage_stat')
 
 local ok = pcall(require, 'ddl')
 if not ok then
@@ -33,6 +34,13 @@ pgroup.before_all(function(g)
     ]])
     t.assert_equals(type(result), 'table')
     t.assert_equals(err, nil)
+
+    helpers.call_on_storages(g.cluster, function(server)
+        server.net_box:eval([[
+            local storage_stat = require('test.helpers.storage_stat')
+            storage_stat.init_on_storage()
+        ]])
+    end)
 end)
 
 pgroup.after_all(function(g) helpers.stop_cluster(g.cluster) end)
@@ -372,70 +380,49 @@ pgroup.test_incomplete_sharding_key_update = function(g)
     t.assert_equals(result, nil)
 end
 
--- Right now CRUD's plan for select doesn't support sharding key and it leads
--- to map reduce (select on all replicasets). To avoid map-reduce one need to
--- add a separate index by field name, used in select's condition. We plan to
--- fix this in scope of https://github.com/tarantool/crud/issues/213
+-- CRUD's plan for select doesn't support sharding key and it leads to map
+-- reduce (select on all replicasets). To avoid map-reduce one need to add a
+-- separate index by field name, used in select's condition. We plan to fix
+-- this in scope of https://github.com/tarantool/crud/issues/213
 pgroup.test_select_wont_lead_map_reduce = function(g)
     local space_name = 'customers_name_key_uniq_index'
     local customers = helpers.insert_objects(g, space_name, {
+        -- bucket_id is 477, storage is s-2
         {id = 1, name = 'Viktor Pelevin', age = 58},
+        -- bucket_id is 401, storage is s-1
         {id = 2, name = 'Isaac Asimov', age = 72},
+        -- bucket_id is 2804, storage is s-2
         {id = 3, name = 'Aleksandr Solzhenitsyn', age = 89},
+        -- bucket_id is 1161, storage is s-2
         {id = 4, name = 'James Joyce', age = 59},
-        {id = 5, name = 'Oscar Wilde', age = 46},
-        {id = 6, name = 'Ivan Bunin', age = 83},
-        {id = 7, name = 'Ivan Turgenev', age = 64},
-        {id = 8, name = 'Alexander Ostrovsky', age = 63},
-        {id = 9, name = 'Anton Chekhov', age = 44},
     })
-    t.assert_equals(#customers, 9)
-
-    -- Disable vshard's rebalancer and account current statistics of SELECT
-    -- calls on storages before calling CRUD select. Rebalancer may screw up
-    -- statistics of SELECT calls, so we will disable it.
-    local servers = g.cluster.servers
-    local select_total_counter_before = 0
-    for n, _ in ipairs(servers) do
-        local c = g.cluster.servers[n].net_box:eval([[
-            local vshard = require('vshard')
-            vshard.storage.rebalancer_disable()
-            assert(vshard.storage.sync(2) == true)
-            assert(vshard.storage.rebalancing_is_in_progress() == false)
-
-            return box.stat().SELECT.total
-        ]])
-        select_total_counter_before = select_total_counter_before + c
-    end
+    t.assert_equals(#customers, 4)
+
+    table.sort(customers, function(obj1, obj2) return obj1.id < obj2.id end)
+
+    local stat_a = storage_stat.collect(g.cluster)
 
     -- Make a CRUD's SELECT.
     local result, err = g.cluster.main_server.net_box:call('crud.select', {
-        space_name, {{'==', 'name', 'Anton Chekhov'}}
+        space_name, {{'==', 'name', 'Viktor Pelevin'}}
     })
     t.assert_equals(err, nil)
     t.assert_not_equals(result, nil)
     t.assert_equals(#result.rows, 1)
 
-    -- Enable vshard's rebalancer and account current statistics of SELECT
-    -- calls on storages after calling CRUD select.
-    local select_total_counter_after = 0
-    for n, _ in ipairs(servers) do
-        local c = g.cluster.servers[n].net_box:eval([[
-            local vshard = require('vshard')
-            local stat = box.stat().SELECT.total
-            vshard.storage.rebalancer_enable()
-
-            return stat
-        ]])
-        select_total_counter_after = select_total_counter_after + c
-    end
+    local stat_b = storage_stat.collect(g.cluster)
 
     -- Compare total counters of SELECT calls on cluster's storages before and
-    -- after calling SELECT on router. Make sure no more than 1 storage changed
+    -- after calling SELECT on a router. Make sure no more than 1 storage changed
     -- SELECT counter. Otherwise we lead map reduce.
-    local diff = select_total_counter_after - select_total_counter_before
-    t.assert_le(diff, 4)
-    t.assert_ge(diff, 2)
+    t.assert_equals(storage_stat.diff(stat_b, stat_a), {
+        ['s-1'] = {
+            select_requests = 0,
+        },
+        ['s-2'] = {
+            select_requests = 1,
+        },
+    })
 end
 
 pgroup.test_select_secondary_idx = function(g)
```

Reply:
Added a note to commit message and a comment to test regarding map-reduce.